File: go/poller.go
package main

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strconv"
	"sync"
	"sync/atomic"
	"time"
)

const (
	maxResponseBody         = 1 * 1024 * 1024
	checkpointIntervalPolls = 10
	pollerMsPerSecond       = 1000
)

type DeviceHealth struct {
	IP                  string
	Label               string
	Status              string
	LastErrorMessage    string
	LastSuccess         int64
	LastError           int64
	ConsecutiveFailures int
}

type Poller struct {
	db        *DB
	cfg       *Config
	health    []DeviceHealth
	successes int64
	failures  int64
	mu        sync.RWMutex
}

func NewPoller(db *DB, cfg *Config) *Poller {
	health := make([]DeviceHealth, len(cfg.Devices))
	for i, d := range cfg.Devices {
		health[i] = DeviceHealth{
			IP:     d.IP,
			Label:  d.Label,
			Status: "unknown",
		}
	}
	return &Poller{db: db, cfg: cfg, health: health}
}

func (p *Poller) Run(ctx context.Context) {
	logf("[poller] Starting — polling %d devices every %.1fs",
		len(p.cfg.Devices), float64(p.cfg.PollIntervalMs)/pollerMsPerSecond)

	p.pollAll()

	ticker := time.NewTicker(time.Duration(p.cfg.PollIntervalMs) * time.Millisecond)
	defer ticker.Stop()

	var pollCount uint32
	for {
		select {
		case <-ctx.Done():
			logf("[poller] Stopped")
			return
		case <-ticker.C:
			p.pollAll()
			pollCount++
			if pollCount%checkpointIntervalPolls == 0 {
				if err := p.db.Checkpoint(); err != nil {
					logf("[poller] checkpoint error: %v", err)
				}
			}
		}
	}
}

func (p *Poller) pollAll() {
	for i := range p.cfg.Devices {
		p.fetchDevice(i)
	}
}

func (p *Poller) fetchDevice(idx int) {
	d := p.cfg.Devices[idx]

	client := &http.Client{
		Timeout: time.Duration(p.cfg.FetchTimeoutMs) * time.Millisecond,
	}

	reqURL := fmt.Sprintf("http://%s/measures/current", d.IP)
	req, err := http.NewRequestWithContext(context.Background(), http.MethodGet, reqURL, http.NoBody)
	if err != nil {
		p.setError(idx, fmt.Sprintf("create request: %v", err))
		return
	}
	resp, err := client.Do(req)
	if err != nil {
		p.setError(idx, fmt.Sprintf("fetch failed: %v", err))
		return
	}
	defer resp.Body.Close()

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		p.setError(idx, fmt.Sprintf("HTTP %d", resp.StatusCode))
		return
	}

	body, err := io.ReadAll(io.LimitReader(resp.Body, maxResponseBody+1))
	if err != nil {
		p.setError(idx, fmt.Sprintf("read body: %v", err))
		return
	}
	if len(body) > maxResponseBody {
		p.setError(idx, "response too large")
		return
	}

	var data map[string]any
	if err := json.Unmarshal(body, &data); err != nil {
		p.setError(idx, "JSON parse error")
		return
	}

	if err := p.db.InsertReading(d.IP, data); err != nil {
		p.setError(idx, fmt.Sprintf("DB insert failed: %v", err))
		return
	}

	logInsertResult(d.Label, d.IP, data)

	atomic.AddInt64(&p.successes, 1)
	p.mu.Lock()
	p.health[idx].Status = "ok"
	p.health[idx].LastSuccess = NowMillis()
	p.health[idx].LastErrorMessage = ""
	p.health[idx].ConsecutiveFailures = 0
	p.mu.Unlock()
}

func (p *Poller) setError(idx int, msg string) {
	d := p.cfg.Devices[idx]
	logf("[poller] %s (%s): %s", d.Label, d.IP, msg)

	atomic.AddInt64(&p.failures, 1)
	p.mu.Lock()
	p.health[idx].Status = "error"
	p.health[idx].LastError = NowMillis()
	p.health[idx].LastErrorMessage = msg
	p.health[idx].ConsecutiveFailures++
	p.mu.Unlock()
}

func (p *Poller) HealthJSON() []map[string]any {
	p.mu.RLock()
	defer p.mu.RUnlock()

	result := make([]map[string]any, len(p.health))
	for i, h := range p.health {
		obj := map[string]any{
			"ip":                  h.IP,
			"label":               h.Label,
			"status":              h.Status,
			"consecutiveFailures": h.ConsecutiveFailures,
		}

		if h.LastSuccess != 0 {
			obj["lastSuccess"] = h.LastSuccess
		} else {
			obj["lastSuccess"] = nil
		}

		if h.LastError != 0 {
			obj["lastError"] = h.LastError
		} else {
			obj["lastError"] = nil
		}

		if h.LastErrorMessage != "" {
			obj["lastErrorMessage"] = h.LastErrorMessage
		} else {
			obj["lastErrorMessage"] = nil
		}

		result[i] = obj
	}
	return result
}

// HealthJSONBytes returns the health status as pre-serialized JSON bytes,
// bypassing encoding/json for maximum throughput.
func (p *Poller) HealthJSONBytes() []byte {
	p.mu.RLock()
	defer p.mu.RUnlock()

	buf := bytes.NewBuffer(make([]byte, 0, len(p.health)*200+2))
	buf.WriteByte('[')
	for i, h := range p.health {
		if i > 0 {
			buf.WriteByte(',')
		}
		buf.WriteString(`{"ip":`)
		writeJSONString(buf, h.IP)
		buf.WriteString(`,"label":`)
		writeJSONString(buf, h.Label)
		buf.WriteString(`,"status":`)
		writeJSONString(buf, h.Status)
		buf.WriteString(`,"lastSuccess":`)
		if h.LastSuccess != 0 {
			buf.WriteString(strconv.FormatInt(h.LastSuccess, 10))
		} else {
			buf.WriteString("null")
		}
		buf.WriteString(`,"lastError":`)
		if h.LastError != 0 {
			buf.WriteString(strconv.FormatInt(h.LastError, 10))
		} else {
			buf.WriteString("null")
		}
		buf.WriteString(`,"lastErrorMessage":`)
		if h.LastErrorMessage != "" {
			writeJSONString(buf, h.LastErrorMessage)
		} else {
			buf.WriteString("null")
		}
		buf.WriteString(`,"consecutiveFailures":`)
		buf.WriteString(strconv.Itoa(h.ConsecutiveFailures))
		buf.WriteByte('}')
	}
	buf.WriteByte(']')
	return buf.Bytes()
}

func (p *Poller) PollStats() (successes, failures int64) {
	return atomic.LoadInt64(&p.successes), atomic.LoadInt64(&p.failures)
}

File: haskell/src/Poller.hs
{-# LANGUAGE OverloadedStrings #-}

module Poller
  ( PollerHandle
  , startPoller
  , getHealthJSON
  , getPollStats
  ) where

import Control.Concurrent (forkIO, threadDelay)
import Control.Concurrent.MVar (MVar, withMVar, newMVar, modifyMVar_)
import Control.Exception (SomeException, try)
import Data.Aeson (Value(..), object, (.=), eitherDecodeStrict')
import Data.IORef (IORef, atomicModifyIORef', newIORef, readIORef)
import Data.Int (Int64)
import System.IO (hPutStrLn, stderr)

import qualified Data.ByteString as BS
import qualified Data.ByteString.Char8 as BS8
import qualified Data.Text as T
import qualified Network.Socket as NS
import qualified Network.Socket.ByteString as NSB

import Config (Config(..), DeviceConfig(..))
import DB (DBHandle, insertReading, checkpoint, nowMillis)

data DeviceHealth = DeviceHealth
  { dhIp                  :: !T.Text
  , dhLabel               :: !T.Text
  , dhStatus              :: !T.Text
  , dhLastSuccess         :: !(Maybe Int64)
  , dhLastError           :: !(Maybe Int64)
  , dhLastErrorMessage    :: !(Maybe T.Text)
  , dhConsecutiveFailures :: !Int
  } deriving (Show)

data PollerHandle = PollerHandle
  { phHealth    :: !(MVar [DeviceHealth])
  , phSuccesses :: !(IORef Int64)
  , phFailures  :: !(IORef Int64)
  }

maxResponseBody :: Int
maxResponseBody = 1 * 1024 * 1024

checkpointIntervalPolls :: Int
checkpointIntervalPolls = 10

startPoller :: MVar DBHandle -> Config -> IO PollerHandle
startPoller dbMVar cfg = do
  let initialHealth =
        [ DeviceHealth
            { dhIp = deviceIp d
            , dhLabel = deviceLabel d
            , dhStatus = "unknown"
            , dhLastSuccess = Nothing
            , dhLastError = Nothing
            , dhLastErrorMessage = Nothing
            , dhConsecutiveFailures = 0
            }
        | d <- cfgDevices cfg
        ]
  healthMVar <- newMVar initialHealth
  succRef <- newIORef 0
  failRef <- newIORef 0
  let h = PollerHandle
        { phHealth = healthMVar
        , phSuccesses = succRef
        , phFailures = failRef
        }
  _ <- forkIO (pollerLoop dbMVar cfg h)
  hPutStrLn stderr $ "[poller] Starting — polling "
    ++ show (length (cfgDevices cfg)) ++ " devices every "
    ++ show (fromIntegral (cfgPollIntervalMs cfg) / 1000 :: Double) ++ "s"
  pure h

pollerLoop :: MVar DBHandle -> Config -> PollerHandle -> IO ()
pollerLoop dbMVar cfg h = do
  pollAll dbMVar cfg h
  go 1
  where
    intervalUs = cfgPollIntervalMs cfg * 1000
    go :: Int -> IO ()
    go pollCount = do
      threadDelay intervalUs
      pollAll dbMVar cfg h
      let newCount = pollCount + 1
      if newCount `mod` checkpointIntervalPolls == 0
        then do
          result <- try $ withMVar dbMVar $ \db -> checkpoint db
          case result of
            Left (e :: SomeException) ->
              hPutStrLn stderr $ "[poller] checkpoint error: " ++ show e
            Right () -> pure ()
        else pure ()
      go newCount

pollAll :: MVar DBHandle -> Config -> PollerHandle -> IO ()
pollAll dbMVar cfg h =
  mapM_ (fetchDevice dbMVar cfg h) (zip [0..] (cfgDevices cfg))

fetchDevice :: MVar DBHandle -> Config -> PollerHandle -> (Int, DeviceConfig) -> IO ()
fetchDevice dbMVar cfg h (idx, dev) = do
  result <- try (fetchHTTP (T.unpack (deviceIp dev)) (cfgFetchTimeoutMs cfg))
  case result of
    Left (e :: SomeException) -> do
      setError h idx dev (show e)
    Right bodyBS -> do
      case eitherDecodeStrict' bodyBS of
        Left err -> setError h idx dev ("JSON parse error: " ++ err)
        Right val -> case val of
          Object _ -> do
            insertResult <- try $ withMVar dbMVar $ \db ->
              insertReading db (deviceIp dev) val
            case insertResult of
              Left (e :: SomeException) ->
                setError h idx dev ("DB insert failed: " ++ show e)
              Right () -> do
                setSuccess h idx
                logPoll dev bodyBS
          _ -> setError h idx dev "Non-object JSON response"

setSuccess :: PollerHandle -> Int -> IO ()
setSuccess h idx = do
  now <- nowMillis
  atomicModifyIORef' (phSuccesses h) (\n -> (n + 1, ()))
  modifyMVar_ (phHealth h) $ \hs ->
    pure (updateAt idx hs $ \dh -> dh
      { dhStatus = "ok"
      , dhLastSuccess = Just now
      , dhLastErrorMessage = Nothing
      , dhConsecutiveFailures = 0
      })

setError :: PollerHandle -> Int -> DeviceConfig -> String -> IO ()
setError h idx dev msg = do
  now <- nowMillis
  hPutStrLn stderr $ "[poller] " ++ T.unpack (deviceLabel dev)
    ++ " (" ++ T.unpack (deviceIp dev) ++ "): " ++ msg
  atomicModifyIORef' (phFailures h) (\n -> (n + 1, ()))
  modifyMVar_ (phHealth h) $ \hs ->
    pure (updateAt idx hs $ \dh -> dh
      { dhStatus = "error"
      , dhLastError = Just now
      , dhLastErrorMessage = Just (T.pack msg)
      , dhConsecutiveFailures = dhConsecutiveFailures dh + 1
      })

updateAt :: Int -> [a] -> (a -> a) -> [a]
updateAt _ [] _ = []
updateAt 0 (x:xs) f = f x : xs
updateAt n (x:xs) f = x : updateAt (n - 1) xs f

getHealthJSON :: PollerHandle -> IO [Value]
getHealthJSON h = do
  hs <- withMVar (phHealth h) pure
  pure (map healthToJSON hs)

healthToJSON :: DeviceHealth -> Value
healthToJSON dh = object
  [ "ip"                  .= dhIp dh
  , "label"               .= dhLabel dh
  , "status"              .= dhStatus dh
  , "lastSuccess"         .= dhLastSuccess dh
  , "lastError"           .= dhLastError dh
  , "lastErrorMessage"    .= dhLastErrorMessage dh
  , "consecutiveFailures" .= dhConsecutiveFailures dh
  ]

getPollStats :: PollerHandle -> IO (Int64, Int64)
getPollStats h = do
  s <- readIORef (phSuccesses h)
  f <- readIORef (phFailures h)
  pure (s, f)

logPoll :: DeviceConfig -> BS.ByteString -> IO ()
logPoll dev _ =
  hPutStrLn stderr $ "[poller] " ++ T.unpack (deviceLabel dev)
    ++ " (" ++ T.unpack (deviceIp dev) ++ "): ok"

-- | Minimal HTTP client using raw sockets with timeout.
fetchHTTP :: String -> Int -> IO BS.ByteString
fetchHTTP ip timeoutMs = do
  let hints = NS.defaultHints { NS.addrSocketType = NS.Stream }
  addrs <- NS.getAddrInfo (Just hints) (Just ip) (Just "80")
  case addrs of
    [] -> fail $ "Cannot resolve: " ++ ip
    (addr:_) -> do
      sock <- NS.openSocket addr
      NS.setSocketOption sock NS.NoDelay 1
      let timeoutUs = timeoutMs * 1000
      NS.setSocketOption sock NS.RecvTimeOut timeoutUs
      NS.setSocketOption sock NS.SendTimeOut timeoutUs
      result <- try $ do
        NS.connect sock (NS.addrAddress addr)
        let req = BS8.pack $ "GET /measures/current HTTP/1.0\r\nHost: "
                           ++ ip ++ "\r\nConnection: close\r\n\r\n"
        sendAllBS sock req
        recvAll sock 0 BS.empty
      NS.close sock
      case result of
        Left (e :: SomeException) -> fail (show e)
        Right body -> extractBody body

sendAllBS :: NS.Socket -> BS.ByteString -> IO ()
sendAllBS _ bs | BS.null bs = pure ()
sendAllBS sock bs = do
  sent <- NSB.send sock bs
  if sent > 0 && sent < BS.length bs
    then sendAllBS sock (BS.drop sent bs)
    else pure ()

recvAll :: NS.Socket -> Int -> BS.ByteString -> IO BS.ByteString
recvAll sock total acc
  | total > maxResponseBody = fail "Response too large"
  | otherwise = do
      chunk <- NSB.recv sock 4096
      if BS.null chunk
        then pure acc
        else recvAll sock (total + BS.length chunk) (acc <> chunk)

extractBody :: BS.ByteString -> IO BS.ByteString
extractBody raw =
  let marker = BS8.pack "\r\n\r\n"
  in case BS8.breakSubstring marker raw of
       (_, after) | BS.null after -> fail "No HTTP body found"
       (hdr, after) ->
         let statusLine = BS8.takeWhile (/= '\r') hdr
             parts = BS8.words statusLine
         in case parts of
              (_:code:_) ->
                let codeStr = BS8.unpack code
                in case codeStr of
                     "200" -> pure (BS.drop 4 after)
                     _     -> fail $ "HTTP " ++ codeStr
              _ -> fail "Malformed HTTP response"
